SLEP 014 Pandas in Pandas out (#37)
scikit-learn/enhancement_proposals: slep014/proposal.rst

Solution
########

The pandas ``DataFrame`` has been widely adopted by the Python Data ecosystem to
store data with feature names. This SLEP proposes using a ``DataFrame`` to
track the feature names as the data is transformed. With this feature, the
API for extracting feature names would be::

    from sklearn import set_config
    set_config(pandas_inout=True)

    pipe.fit(X, y)
    X_trans = pipe[:-1].transform(X)

    print(X_trans.columns.tolist()
    ['letter_a', 'letter_b', 'letter_c', 'pet_dog', 'pet_snake', 'num']

Enabling Functionality
######################

The following enhancements are **not** a part of this SLEP. These features are
made possible if this SLEP gets accepted.

1. Allows estimators to treat columns differently based on name or dtype. For
   example, the categorical dtype is useful for tree building algorithms.

2. Storing feature names inside estimators for model inspection::

    from sklearn import set_config
    set_config(store_feature_names_in=True)

    pipe.fit(X, y)

    pipe['logisticregression'].feature_names_in_

3. Allow for extracting the feature names of estimators in meta-estimators::

    from sklearn import set_config

>     set_config(store_feature_names_in=True)

Author: The default of `pandas_inout` is now stated twice in the Solution and Backward compatibility sections. The default value of `store_feature_names_in` is started in its section. I purposefully did not go into too many details in the **Enabling Functionality** section, since it serves as "what are the possibilities if this SLEP gets accepted".

        return construct_result(result, output_feature_names, row_labels, input_type)

Memory copies
-------------

As noted in `pandas #27211 <https://github.com/pandas-dev/pandas/issues/27211>`_,
there is not a guarantee that there is a zero-copy round-trip going from numpy
to a ``DataFrame``. In other words, the following may lead to a memory copy in
a future version of ``pandas``::

    X = np.array(...)
    X_df = pd.DataFrame(X)
    X_again = np.asarray(X_df)

This is an issue for ``scikit-learn`` when estimators are placed into a
pipeline. For example, consider the following pipeline::

    set_config(pandas_inout=True)
    pipe = make_pipeline(StandardScaler(), LogisticRegression())
    pipe.fit(X, y)

Interally, ``StandardScaler.fit_transform`` will operate on a ndarray and
wrap the ndarray into a ``DataFrame`` as a return value. This is will be
piped into ``LogisticRegression.fit`` which calls ``check_array`` on the
``DataFrame``, which may lead to a memory copy in a future version of
``pandas``. This leads to unnecessary overhead from piping the data from one
estimator to another.

Backward compatibility
######################

> The ``set_config(pandas_inout=True)`` global configuration flag will be set to

Reviewer: maybe more like `pandas_output`?

Author: I had something like that initially, but it feels like "We will always output pandas, even if the input is numpy arrays". (Naming is hard)

    cat_names = (pipe['columntransformer']
                 .named_transformers_['onehotencoder']
                 .get_feature_names(orig_cat_cols))

    feature_names = np.r_[cat_names, orig_num_cols]

The ``feature_names`` extracted above corresponds to the features directly
passed into ``LogisticRegression``. As demonstrated above, the process of
extracting ``feature_names`` requires knowing the order of the selected
categories in the ``ColumnTransformer``. Furthemore, if there is feature
selection in the pipeline, such as ``SelectKBest``, the ``get_support`` method
would need to be used to select column names that were selected.

Solution
########

The pandas DataFrame has been widely adopted by the Python Data ecosystem to
store data with feature names. This SLEP proposes using a DataFrame to
track the feature names as the data is transformed. With this feature, the
API for extracting feature names would be::

    from sklearn import set_config
    set_config(pandas_inout=True)

    pipe.fit(X, y)
    X_trans = pipe[:-1].transform(X)

    X_trans.columns.tolist()
    ['letter_a', 'letter_b', 'letter_c', 'pet_dog', 'pet_snake', 'num']

> This introduces a soft dependency on pandas, which is opt-in with the

Reviewer: So maybe word it more like

> This feature is only available through a soft dependency on `pandas`

or something.

the configuration flag: ``pandas_inout``. By default, ``pandas_inout`` is set
to ``False``, resulting in the output of all estimators to be a ndarray.

Enabling Functionality
######################

The following enhancements are **not** a part of this SLEP. These features are
made possible if this SLEP gets accepted.

1. Allows estimators to treat columns differently based on name or dtype. For
   example, the categorical dtype is useful for tree building algorithms.

2. Storing feature names inside estimators for model inspection::

    from sklearn import set_config
    set_config(store_feature_names_in=True)

    pipe.fit(X, y)

    pipe['logisticregression'].feature_names_in_

3. Allow for extracting the feature names of estimators in meta-estimators::

    from sklearn import set_config
    set_config(store_feature_names_in=True)

    est = BaggingClassifier(LogisticRegression())
    est.fit(X, y)

    # Gets the feature names used by an estimator in the ensemble
    est.estimators_[0].feature_names_in_

For options 2 and 3 the default value of configuration flag:
`store_feature_names_in` is False.

Considerations
##############

Index alignment
---------------

> Operations are index aligned when working with DataFrames. Interally,

Reviewer 2: ```suggestion
Operations are index aligned when working with DataFrames. Internally,
```

        # preserving row labels, attaching new features names.
        return construct_result(result, output_feature_names, row_labels, input_type)

Memory copies
-------------

As noted in `pandas #27211 <https://github.com/pandas-dev/pandas/issues/27211>`_,
there is not a guarantee that there is a zero-copy round-trip going from numpy
to a DataFrame. In other words, the following may lead to a memory copy in
a future version of ``pandas``::

    X = np.array(...)
    X_df = pd.DataFrame(X)
    X_again = np.asarray(X_df)

This is an issue for ``scikit-learn`` when estimators are placed into a
pipeline. For example, consider the following pipeline::

    set_config(pandas_inout=True)
    pipe = make_pipeline(StandardScaler(), LogisticRegression())
    pipe.fit(X, y)

Interally, ``StandardScaler.fit_transform`` will operate on a ndarray and
wrap the ndarray into a DataFrame as a return value. This is will be
piped into ``LogisticRegression.fit`` which calls ``check_array`` on the
DataFrame, which may lead to a memory copy in a future version of
``pandas``. This leads to unnecessary overhead from piping the data from one

> estimator to another.

Reviewer 2: Personally, I think that for some transformers (like StandardScaler) could rather easily work column-wise to avoid such copying overhead. 
(of course, that will give a stronger pandas dependence also for the implementation of such transformers)

Reviewer: and it could even have the option of being "in-place" :D

Author: We can try to support this. `check_array` would need to not run `asarray` on the dataframe, and the transformer would need to operate on the dataframe itself.

Reviewer 3: yes ``check_array`` could get another option! exciting! ;)

    pipe.fit(X, y)

Interally, ``StandardScaler.fit_transform`` will operate on a ndarray and
wrap the ndarray into a DataFrame as a return value. This is will be
piped into ``LogisticRegression.fit`` which calls ``check_array`` on the
DataFrame, which may lead to a memory copy in a future version of
``pandas``. This leads to unnecessary overhead from piping the data from one
estimator to another.

Backward compatibility
######################

The ``set_config(pandas_in_out=True)`` global configuration flag will be set to
``False`` by default to ensure backward compatibility. When this flag is False,
the output of all estimators will be a ndarray.

Alternatives
############

- :ref:`SLEP012 Custom InputArray Data Structure <slep_012>`: This approach
  adds another data structure in the Python Data ecosystem. This increases
  the maintenance responsibilities of the ``scikit-learn`` library.

- Use xarray's Dataset, ``xr.Dataset``: The pandas DataFrame is more widely used
  in Python's Data ecosystem, which means more libraries are built with pandas
  in mind. With xarray support, users will need to convert their DataFrame into

>   a ``xr.Dataset``. This converstion process will be lossy when working with

Reviewer 4: ```suggestion
  a ``xr.Dataset``. This conversion process will be lossy when working with
```

    from sklearn.preprocessing import OneHotEncoder, StandardScaler
    from sklearn.pipeline import make_pipeline
    from sklearn.linear_model import LogisticRegression

    X = pd.DataFrame({'letter': ['a', 'b', 'c'], 
                      'pet': ['dog', 'snake', 'dog'],
                      'num': [1, 2, 3]})
    y = [0, 0, 1]
    orig_cat_cols, orig_num_cols = ['letter', 'pet'], ['num']

    ct = make_column_transformer(
        (OneHotEncoder(), orig_cat_cols), (StandardScaler(), orig_num_cols))
    pipe = make_pipeline(ct, LogisticRegression()).fit(X, y)

    cat_names = (pipe['columntransformer']
                 .named_transformers_['onehotencoder']
                 .get_feature_names(orig_cat_cols))

    feature_names = np.r_[cat_names, orig_num_cols]

The ``feature_names`` extracted above corresponds to the features directly
passed into ``LogisticRegression``. As demonstrated above, the process of
extracting ``feature_names`` requires knowing the order of the selected
categories in the ``ColumnTransformer``. Furthemore, if there is feature
selection in the pipeline, such as ``SelectKBest``, the ``get_support`` method

> would need to be used to select column names that were selected.

Reviewer 5: ```suggestion
would need to be used to determine column names that were selected.
```

Solution
########

The pandas DataFrame has been widely adopted by the Python Data ecosystem to
store data with feature names. This SLEP proposes using a DataFrame to
track the feature names as the data is transformed. With this feature, the
API for extracting feature names would be::

    from sklearn import set_config
    set_config(pandas_in_out=True)

    pipe.fit(X, y)
    X_trans = pipe[:-1].transform(X)

    X_trans.columns.tolist()
    ['letter_a', 'letter_b', 'letter_c', 'pet_dog', 'pet_snake', 'num']

This SLEP proposes attaching feature names to the output of ``transform``. In
the above example, ``pipe[:-1].transform(X)`` propagates the feature names
through the multiple transformers.

This feature is only available through a soft dependency on pandas. Furthermore,
it will be opt-in with the the configuration flag: ``pandas_in_out``. By
default, ``pandas_in_out`` is set to ``False``, resulting in the output of all
estimators to be a ndarray.

> 

Reviewer 5: It's not clear here whether pandas output will be provided even when the input is ndarray.

I don't see why that should only be a matter of "community adoption".

Enabling Functionality
######################

The following enhancements are **not** a part of this SLEP. These features are
made possible if this SLEP gets accepted.

> 1. Allows estimators to treat columns differently based on name or dtype. For

Reviewer 5: In what way is this enabled by the present SLEP? I assume this means something more expansive: that we will try to retain dtype in outputting a dataframe e.g. after feature selection.

Otherwise this pertains to the handling of Pandas input, which is done on a case-by-case basis already?

a future version of ``pandas``::

    X = np.array(...)
    X_df = pd.DataFrame(X)
    X_again = np.asarray(X_df)

This is an issue for ``scikit-learn`` when estimators are placed into a
pipeline. For example, consider the following pipeline::

    set_config(pandas_in_out=True)
    pipe = make_pipeline(StandardScaler(), LogisticRegression())
    pipe.fit(X, y)

Interally, ``StandardScaler.fit_transform`` will operate on a ndarray and
wrap the ndarray into a DataFrame as a return value. This is will be
piped into ``LogisticRegression.fit`` which calls ``check_array`` on the
DataFrame, which may lead to a memory copy in a future version of
``pandas``. This leads to unnecessary overhead from piping the data from one
estimator to another.

Sparse matrices
---------------

Traditionally, ``scikit-learn`` prefers to process sparse matrices in
the compressed sparse row (CSR) matrix format. The `sparse data structure <https://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html>`_ in pandas 1.0 only supports converting directly to
the coordinate format (COO). Although this format was designed to quickly
convert to CSR or CSC formats, the conversion process still needs to allocate

> more memory to store. This can be an issue with transformers such as the

Reviewer 6: This seems to be a fairly conservative statement. Last time I tried to create pandas sparse frame  with 100k features, a year or so ago, that was very slow. We should check whether it's realistic to use pandas sparse format for current scikit-learn use cases.

``OneHotEncoder.transform`` which has been optimized to construct a CSR matrix.

> 

Reviewer 5: Estimators without rectangular input: should they produce pandas output, e.g. how do you get a vectorizer to produce pandas out?

The ``set_config(pandas_in_out=True)`` global configuration flag will be set to
``False`` by default to ensure backward compatibility. When this flag is False,
the output of all estimators will be a ndarray.

Community Adoption
##################

With the new ``pandas_in_out`` configuration flag, third party libraries may
need to query the configuration flag to be fully compliant with this SLEP.
Specifically, "to be fully compliant" entails the following policy:

1. If ``pandas_in_out=False``, then ``transform`` always returns numpy array.
2. If ``pandas_in_out=True``, then ``transform`` returns a DataFrame if the
   input is a Dataframe.

This policy can either be enforced with ``check_estimator`` or not:

- **Enforce**: This increases the maintaince burden of third party libraries.
  This burden includes: checking for the configuration flag, generating feature names and including pandas as a dependency to their library.

- **Not Enforce**: Currently, third party transformers can return a DataFrame
  or a numpy and this is mostly compatible with ``scikit-learn``. Users with
  third party transformers would not be able to access the features enabled
  by this SLEP.

> 

Reviewer 5: Also: code relying on the output of Scikit-learn transformers will presume it is an ndarray presently, but will get something different now.

Alternatives
############

This section lists alternative data structures that can be used with their
advantages and disadvantages when compared to a pandas DataFrame.

InputArray
----------

The proposed ``InputArray`` described
:ref:`SLEP012 Custom InputArray Data Structure <slep_012>` introduces a new
data structure for homogenous data.

Pros
~~~~

- A thin wrapper around a numpy array or a sparse matrix with a minimial feature
  set that ``scikit-learn`` can evolve independently.

Cons
~~~~

- Introduces another data structure for data storage in the PyData ecosystem.
- Currently, the design only allows for homogenous data.
- Increases maintenance responsibilities for ``scikit-learn``.

XArray Dataset
--------------

`xarray's Dataset <http://xarray.pydata.org/en/stable/data-structures.html#dataset>`_
is a multi-dimenstional version of panda's DataFrame.

Pros
~~~~

> - Can be used for heterogeneous data.

Reviewer 6: I would say that handling of sparse is a disadvantage in both,
- xarray supports https://github.com/pydata/sparse as far as I understand, but whether we want to use it and what is the conversion cost remains to be evaluated
- conversion back and forth to pandas sparse frame also has non negligible costs.

default, ``pandas_in_out`` is set to ``False``, resulting in the output of all
estimators to be a ndarray.

Enabling Functionality
######################

The following enhancements are **not** a part of this SLEP. These features are
made possible if this SLEP gets accepted.

1. Allows estimators to treat columns differently based on name or dtype. For
   example, the categorical dtype is useful for tree building algorithms.

2. Storing feature names inside estimators for model inspection::

    from sklearn import set_config
    set_config(store_feature_names_in=True)

    pipe.fit(X, y)

    pipe['logisticregression'].feature_names_in_

3. Allow for extracting the feature names of estimators in meta-estimators::

    from sklearn import set_config
    set_config(store_feature_names_in=True)

    est = BaggingClassifier(LogisticRegression())
    est.fit(X, y)

    # Gets the feature names used by an estimator in the ensemble
    est.estimators_[0].feature_names_in_

For options 2 and 3 the default value of configuration flag:
``store_feature_names_in`` is False.

Considerations
##############

Index alignment
---------------

> Operations are index aligned when working with DataFrames. Internally,

Reviewer 7: It'd be useful to briefly describe what index-alignment means for dataframes, or to provide a link.

For now I share @TomAugspurger opinion and I honestly don't see where index-alignment would ever be a concern for us. So maybe we could just remove the section.

to a DataFrame. In other words, the following may lead to a memory copy in
a future version of ``pandas``::

    X = np.array(...)
    X_df = pd.DataFrame(X)
    X_again = np.asarray(X_df)

This is an issue for ``scikit-learn`` when estimators are placed into a
pipeline. For example, consider the following pipeline::

    set_config(pandas_in_out=True)
    pipe = make_pipeline(StandardScaler(), LogisticRegression())
    pipe.fit(X, y)

Interally, ``StandardScaler.fit_transform`` will operate on a ndarray and
wrap the ndarray into a DataFrame as a return value. This is will be
piped into ``LogisticRegression.fit`` which calls ``check_array`` on the
DataFrame, which may lead to a memory copy in a future version of
``pandas``. This leads to unnecessary overhead from piping the data from one
estimator to another.

Sparse matrices
---------------

Traditionally, ``scikit-learn`` prefers to process sparse matrices in
the compressed sparse row (CSR) matrix format. The `sparse data structure <https://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html>`_ in pandas 1.0, only supports converting directly to
the cooridnate format (COO). Although this format was designed to quickly

> convert to CSR or CSC formats, the converation process still needs to allocate

Reviewer 7: ```suggestion
the compressed sparse row (CSR) matrix format. The `sparse data structure <https://pandas.pydata.org/pandas-docs/stable/user_guide/sparse.html>`_ in pandas 1.0 only supports converting directly to
the coordinate format (COO). Although this format was designed to quickly
convert to CSR or CSC formats, the conversion process still needs to allocate
```

  third party transformers would not be able to access the features enabled
  by this SLEP.

Alternatives
############

This section lists alternative data structures that can be used with their
advantages and disadvantages when compared to a pandas DataFrame.

InputArray
----------

The proposed ``InputArray`` described
:ref:`SLEP012 Custom InputArray Data Structure <slep_012>` introduces a new
data structure for homogenous data.

Pros
~~~~

- A thin wrapper around a numpy array or a sparse matrix with a minimial feature
  set that ``scikit-learn`` can evolve independently.

Cons
~~~~

- Introduces another data structure for data storage in the PyData ecosystem.
- Currently, the design only allows for homogenous data.
- Increases maintenance responsibilities for ``scikit-learn``.

XArray Dataset
--------------

`xarray's Dataset <http://xarray.pydata.org/en/stable/data-structures.html#dataset>`_
is a multi-dimenstional version of panda's DataFrame.

Pros
~~~~

- Can be used for heterogeneous data.

Cons
~~~~

- ``scikit-learn`` does not require many of the features Dataset provides.
- Needs to be converted to a DataArray before it can be converted to a numpy array.

> - The `conversation from a pandas DataFrame to a Dataset <http://xarray.pydata.org/en/stable/pandas.html>`_ 

Reviewer 7: ```suggestion
- The `conversion from a pandas DataFrame to a Dataset <http://xarray.pydata.org/en/stable/pandas.html>`_ 
```

  is not lossless. For example, categorical dtypes in a pandas dataframe will

>   lose its categorical information when converted to a Dataset.

Reviewer 7: ```suggestion
  lose their categorical information when converted to a Dataset.
```

- xarray does not have as much adoption as pandas, which increases the learning
  curve for using Dataset with ``scikit-learn``.

XArray DataArray
----------------

`xarray's DataArray <http://xarray.pydata.org/en/stable/data-structures.html#dataarray>`_
is a data structure that store homogenous data.

Pros
~~~~

> - xarray guartantees that there will be no copies during round-trips from

Reviewer 7: ```suggestion
- xarray guarantees that there will be no copies during round-trips from
```

  numpy. (`xarray #3077 <https://github.com/pydata/xarray/issues/3077>`_)

Cons
~~~~

- Can only be used for homogenous data.

> - As with XArray's Dataset, DataArray does not have much adoption as pandas,

Reviewer 7: ```suggestion
- As with XArray's Dataset, DataArray does not have as much adoption as pandas,
```
